text/0087-graphql-errors.md

```diff
@@ -0,0 +1,228 @@
+
+```json
+{
+  "errors": [
+    {
+      "path": [
+        "query",
+        "viewer",
+        "logi"
+      ],
+      "extensions": {
+        "code": "undefinedField",
+        "typeName": "User",
+        "fieldName": "logi"
+      },
+      "locations": [
+        {
+          "line": 4,
+          "column": 5
+        }
+      ],
+      "message": "Field 'logi' doesn't exist on type 'User'"
+    }
+  ]
+}
+```
+
+Using the `locations` and the request body, we can highlight the error in the request body.
+
+```
+query { 
+  viewer { 
+    # note the removed `n` at the end
+    -> logi
+  }
+}
+```
+
+Because of that, the `Request` and `Response` body should be sent to Sentry.
+
+# Supporting Data
+
+See [JIRA issues](https://www.notion.so/sentry/GraphQL-proposal-d6e5846f30434770903cf3af20bc2568?pvs=4#fd0b946340d947eb9bbaa320846ccd6a), not disclosing here because they could contain PII.
+
+Related issues and discussions:
+
+[GraphQL Support](https://github.com/getsentry/sentry/discussions/38913)
+
+[Support for GraphQL errors](https://github.com/getsentry/sentry/issues/33723)
+
+[Send request body for http client integrations and similar](https://github.com/getsentry/team-mobile/issues/41)
+
+# Proposal (Option 1)
+
+The proposal is adding a `data` field in the [Response](https://develop.sentry.dev/sdk/event-payloads/types/#responsecontext) interface.
+
+By doing this, we can keep the `Request` interface as it is, we can copy the `data` scrubbing rules from the `Request` interface.
+
+```json
+{
+  "contexts": {
+    "response": {
+      "type": "response",
+      "status_code": 500,
+      "body_size": 1000,
+      "data": "...",
+    }
+  }
+}
+```
+
+* `data`: Can be given as string or structural data of any format..
+
+The `Response` interface keeps arbitrary fields, it's back compatible with the current implementation.
+
+## Must have
+
+The fields `Request#data` and `Response#data` could contain PII and they should run data scrubbing agressively.
+
+[Session Replay](https://docs.sentry.io/platforms/javascript/guides/remix/session-replay/configuration/) already sends the request and response bodies, so we can use the same data scrubbing rules.
+
+Since GraphQL is a well defined spec, we can also scrub the GraphQL fields.
+
+Request example:
+
+```
+query { 
+  viewer { 
+    login
+  }
+}
+```
+
+Response example:
+
+```json
+{
+  "data": {
+    "viewer": {
+      "login": "[Filtered]"
+    }
+  }
+}
+```
+
+# Drawbacks
+
+[Envelopes](https://develop.sentry.dev/sdk/envelopes) (Events) contain way lower [size limits](https://develop.sentry.dev/sdk/envelopes/#size-limits). The `data` fields could be large and it could be a problem.
+
+SDKs should discard large and binary bodies by default, using the [maxRequestBodySize](https://docs.sentry.io/platforms/android/configuration/options/#max-request-body-size) and `maxResponseBodySize` (it'll be added) options.
+
+The difference is that for GraphQL errors, this should be enabled by default.
```

Comment:
If we don't send the payload, the user won't have context about what has to be fixed.

Reply:
The other option is to manipulate the request and payload sent, for example, sending only the `errors` of the response and the body without the `variables` from the request, which requires serializing/deserializing the request and response of the client, this might add performance overhead as just sending it as it is, but it's an option.

Reply:
How will we handle situations where partial data are returned with errors? Will we only send the errors to Sentry or the filtered data too?

Reply:
It's still an error, https://www.notion.so/sentry/GraphQL-proposal-d6e5846f30434770903cf3af20bc2568?pvs=4#65c51cb28c1c421e8b3edd4f5fe2f427
I'll check during the implementation if there's a performance impact of serializing the error, extracting the non-required fields, etc.